pkg/controller/admissionchecks/multikueue/controller.go

```diff
@@ -0,0 +1,374 @@
+)
+
+type multiKueueStoreHelper = admissioncheck.ConfigHelper[*kueue.MultiKueueConfig, kueue.MultiKueueConfig]
+
+func newMultiKueueStoreHelper(c client.Client) (*multiKueueStoreHelper, error) {
+	return admissioncheck.NewConfigHelper[*kueue.MultiKueueConfig](c)
+}
+
+type AcReconciler struct {
+	client client.Client
+	helper *multiKueueStoreHelper
+
+	lock        sync.RWMutex
+	controllers map[string]*remoteController
+	wlUpdateCh  chan event.GenericEvent
+
+	rootContext context.Context
+
+	// for testing only
+	updateConfigOverride func(ctx context.Context, rc *remoteController, kubeconfigs map[string][]byte) error
+}
+
+var _ reconcile.Reconciler = (*AcReconciler)(nil)
+var _ manager.Runnable = (*AcReconciler)(nil)
+
+func (a *AcReconciler) controllerFor(acName string) (*remoteController, bool) {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+
+	c, f := a.controllers[acName]
+	return c, f
+}
+
+func (a *AcReconciler) setControllerFor(acName string, c *remoteController) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	if old, found := a.controllers[acName]; found {
+		old.watchCancel()
+	}
+	if c != nil {
+		a.controllers[acName] = c
+	} else {
+		delete(a.controllers, acName)
+	}
+}
+
+func (a *AcReconciler) Start(ctx context.Context) error {
+	a.rootContext = ctx
+	return nil
+}
+
+func (a *AcReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	ac := &kueue.AdmissionCheck{}
+	if err := a.client.Get(ctx, req.NamespacedName, ac); err != nil || ac.Spec.ControllerName != ControllerName {
+		if apierrors.IsNotFound(err) || !ac.DeletionTimestamp.IsZero() {
+			// stop/deleted a potential check controller
+			if cc, existing := a.controllerFor(req.Name); existing {
+				cc.watchCancel()
+				a.setControllerFor(req.Name, nil)
+				log.V(2).Info("Controller removed")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	var remCtrl *remoteController
+	inactiveReason := ""
+
+	log.V(2).Info("Reconcile AdmissionCheck")
+	if cfg, err := a.helper.ConfigFromRef(ctx, ac.Spec.Parameters); err != nil {
+		inactiveReason = fmt.Sprintf("Cannot load AC config: %s", err.Error())
+	} else {
+		kubeconfigs, err := a.getKubeConfigs(ctx, &cfg.Spec)
+		if err != nil {
+			a.setControllerFor(ac.Name, nil)
+			inactiveReason = fmt.Sprintf("Cannot load kubeconfigs: %s", err.Error())
+		} else {
+			cc, existing := a.controllerFor(ac.Name)
+			if !existing {
+				cc = newRemoteController(a.rootContext, a.client, a.wlUpdateCh)
+				a.setControllerFor(ac.Name, cc)
+			}
+
+			var err error
+			if a.updateConfigOverride != nil {
+				err = a.updateConfigOverride(ctx, cc, kubeconfigs)
+			} else {
+				err = cc.UpdateConfig(ctx, kubeconfigs)
+			}
+
+			if err != nil {
+				inactiveReason = fmt.Sprintf("Cannot start remote controller: %s", err.Error())
+				a.setControllerFor(ac.Name, nil)
+			} else {
+				remCtrl = cc
+			}
+		}
+	}
+
+	newCondition := metav1.Condition{
```

Comment:
The KEP mentions a new admission check (called MultiKueue), in the first sentence. Do we manage it here?